### PR TITLE
Kotlin leg tweaks

### PR DESCRIPTION
**Pull request by Stermere**
Converts the legTweaks files to Kotlin and improves the readability a touch. ~~There should be no difference in functionality~~ This fixes a minor bug that has been present since the math lib switch, so there should be a slight improvement!

### CODE REVIEW

File: server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.kt [hunk 1]
```diff
@@ -0,0 +1,1167 @@
...
+	private var bufferInvalid = true
+
+	constructor(skeleton: HumanSkeleton, config: LegTweaksConfig) : this(skeleton) {
+		this.config = config
+		updateConfig()
+	}
+
+	fun getLeftFootPosition(): Vector3 {
+		return leftFootPosition
+	}
```
Comment (Louka3000) [inline server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.kt:132]: I would just make all of these public variables and remove the get/set

File: server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaksBuffer.kt [hunk 2]
```diff
@@ -0,0 +1,821 @@
...
+
+	fun setCorrectedRotations(leftFootRotation: Quaternion, rightFootRotation: Quaternion) {
+		this.leftFootRotationCorrected = leftFootRotation
+		this.rightFootRotationCorrected = rightFootRotation
+	}
+
+	// getters and setters
+	fun getLeftFootPosition(): Vector3 {
+		return leftFootPosition
+	}
```
Comment (Louka3000) [inline server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaksBuffer.kt:242]: Same here, just make them public imo

File: server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaksBuffer.kt [hunk 3]
```diff
@@ -0,0 +1,821 @@
...
+	private fun detectOutsideForces(f1: Vector3, f2: Vector3): Boolean {
+		val force: Vector3 = GRAVITY.plus(f1).plus(f2)
+		val error: Vector3 = centerOfMassAcceleration.minus(force)
+		return error.lenSq() > FORCE_ERROR_TOLERANCE_SQR
+	}
+
+	// simple error function for the force vector gradient descent
+	private fun getForceVectorError(testForce: Vector3, otherForce: Vector3): Vector3 {
+		return centerOfMassAcceleration
+			.minus(testForce.plus(otherForce).plus(GRAVITY))
```
Comment (Louka3000) [inline server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaksBuffer.kt:819]:
```suggestion
		return centerOfMassAcceleration - testForce + otherForce + GRAVITY
```
Just as an example, you can do this in Kotlin, which makes it easier to read IMO. You can look through your code and use +, -, * and / wherever

**Review by Louka3000 — CHANGES_REQUESTED** [review]

File: server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.kt (hunk 1 above)
Comment (Stermere) [reply server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.kt]: didn't know Kotlin had that feature! thanks

File: server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.kt [hunk 4]
```diff
@@ -0,0 +1,1079 @@
...
+
+	fun setSkatingReductionEnabled(skatingCorrectionEnabled: Boolean) {
+		this.skatingCorrectionEnabled = skatingCorrectionEnabled
+
+		// reset the buffer
+		bufferHead = LegTweaksBuffer()
+		bufferInvalid = true
+	}
+
+	fun setToeSnapEnabled(value: Boolean) {
```
Comment (Louka3000) [inline server/core/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.kt:151]: You should also be able to remove these getters and setters up to line 186

**Review by Louka3000 — APPROVED** [review]

**Review by ButterscotchV — APPROVED** [review]
Code looks fine, but I wouldn't trust my review as a definite. It was very hard to review this so I definitely could've missed something... Just make sure this is well tested please. If not before the PR is merged, then *absolutely* before a release is made.

**Comment by Stermere** [comment]
I've tested all the features that I would expect to break but have not yet done all the integration tests. Once I have and everything works as it should I'll update this comment.

**Comment by ImUrX** [comment]
how is this PR doing?

**Comment by Stermere** [comment]
> how is this PR doing?

It's ready, I re-tested it after rebasing it to the bones rewrite about 2 weeks ago.

**Review by ImUrX — APPROVED** [review]

**Pull request merged**